---
title: "A. Data Discovery and Retrieval"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{A. Data Discovery and Retrieval}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

CxGcensus is an alternative *R* client to the [CELLxGENE census][]. It
emaphsizes use cases related to data discovery, and uses in-memory and
on-disk caches to reduce latency associated with repeated queries.

[CELLxGENE census]: https://chanzuckerberg.github.io/cellxgene-census/

# Installation

Install CxGcensus from [GitHub](https://github.com/) with:

``` r
## install.packages("devtools")
devtools::install_github("mtmorgan/CxGcensus")
```

# Data exploration

Load the package

```{r example, message = FALSE}
library(CxGcensus)
```

This is the census and package versions we are using

```{r census}
census()$get_metadata() |>
    lapply(as.vector) |>
    str()
packageVersion("CxGcensus")
tiledbsoma::show_package_versions()
```

Discover datasets used to construct the census

```{r datasets}
datasets()
```

`dataset_id` and `collection_id` can be used in the [cellxgenedp][]
package for programmatic retrieval and visualization of individual
datasets.

The organisms present in the census (use these in calls to
`feature_data()` or `observation_data()`) are available with

```{r}
census_names()
```

Summarize information about cells in the census

```{r summary_cell_counts}
summary_cell_counts() |>
    filter(category == "sex") |>
    select(
        organism, label,
        unique_cell_count, total_cell_count
    )
```

Learn about features (genes) in, e.g., `homo_sapiens` datasets in the census

```{r cars}
feature_data("mus_musculus")
```

Observation (cell) data annotations are available with
`observation_data()`. The strategy adopted is to download all the data
to a local [duckdb][] database. Download can be time-consuming, but
the results are 'memoised' so subsequent calls are more-or-less
instantaneous.

```{r observation_data}
mus <- observation_data("mus_musculus")
mus
```

The columns available for cross-dataset analysis are

```{r}
mus |>
    colnames()
```

It is straight-forward to summarize or query the data, e.g., for the
number of different types of assays represented, or for studies of
diabetes.

```{r}
mus |>
    count(assay, sort = TRUE)
mus |>
    filter(grepl("diabetes", disease)) |>
    count(disease, sex, tissue)
```

Use [ggplot2][] to visualize tissue type and assay.

```{r mus-tissue-assay, message = FALSE}
library(ggplot2)

## 10 most-common tissues; total cell count
common_tissues <-
    mus |>
    count(tissue, sort = TRUE) |>
    head(10) |>
    collect() |>
    pull(tissue)
## assays on each tissue
tissue_and_assay <-
    mus |>
    filter(tissue %in% common_tissues) |>
    count(tissue, assay, sort = TRUE) |>
    collect() |>
    mutate(n = as.integer(n))
## visualize
ggplot(tissue_and_assay) +
    aes(x = factor(tissue, levels = common_tissues), y = n, fill = assay) +
    geom_bar(position="stack", stat="identity") +
    labs(x = "Tissue", y = "Total cell count") +
    coord_flip() +
    theme(legend.position = c(.8, .7))
```

The `soma_joinid` in the tibbles returned by `feature_data()` and
`observation_data()` are important in retrieving counts.

[cellxgenedp]: https://bioconductor.org/packages/cellxgenedp
[duckdb]: https://cran.r-project.org/package=duckdb
[ggplot2]: https://cran.r-project.org/package=ggplot2

# Experiment retrieval as SingleCellExperiment

To create a Bioconductor [SingleCellExperiment][], perhaps for use in
one of the [Orchestrating Single Cell Analysis with
Bioconductor][OSCA] (OSCA) workflows, first select desired features
and observations. For illustration we choose 100 random genes

```{r features}
set.seed(12)
features <- feature_data("mus_musculus")
random_features <-
    features |>
    slice(sample(nrow(features), 100))
```

and choose all cells from brain tissue and development stage 20 month
old or latter.

```{r observations}
observations <- observation_data("mus_musculus")
brain_20mo <-
    observations |>
    filter(
        tissue == "brain",
        development_stage == "20 month-old stage and over"
    ) |>
    collect()
```

Cells are from several datasets

```{r brain-summary}
brain_20mo |>
    count()
brain_20mo |>
    count(dataset_id)
```

Use these to create a `SingleCellExperiment`.

```{r single-cell-experiment}
sce <- single_cell_experiment("mus_musculus", random_features, brain_20mo)
sce
```

The experiment has `counts()` as a sparse (`dgCMatrix`, from the
Matrix package) matrix, with `features` and `observations` available
(as *Bioconductor* `DataFrame` objects, rather than tibbles) via
`rowData()` and `colData()`. The data is straight-forward to work
with, e.g., removing features for which no counts were observed...

```{r sce-summary, message = FALSE}
library(SingleCellExperiment)
keep_rows <- Matrix::rowSums(counts(sce, withDimnames = FALSE)) != 0
table(keep_rows)
sce[keep_rows,]
```

... or summarizing the total number of reads observed in each cell

```{r sce-colSums}
hist(
    log1p(colSums(counts(sce, withDimnames = FALSE))),
    main = "Counts per cell",
    xlab = "log(1 + x) counts",
)
```

[SingleCellExperiment]: https://bioconductor.org/packages/SingleCellExperiment
[OSCA]: https://bioconductor.org/books/OSCA

# Session information

This article was compiled with CxGcensus version
`r packageVersion("CxGcensus")`. Full session info is:

```{r}
sessionInfo()
```
