<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CxGcensus">
<title>A. Data Discovery and Retrieval • CxGcensus</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A. Data Discovery and Retrieval">
<meta property="og:description" content="CxGcensus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CxGcensus</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9025</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/a_discovery_and_retrieval.html">A. Data Discovery and Retrieval</a>
    <a class="dropdown-item" href="../articles/b_integration.html">B. Dataset Integration</a>
    <a class="dropdown-item" href="../articles/c_compare.html">C. Comparison with CuratedAtlasQuery and cellxgene.census</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mtmorgan/CxGcensus">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>A. Data Discovery and Retrieval</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mtmorgan/CxGcensus/blob/HEAD/vignettes/a_discovery_and_retrieval.Rmd" class="external-link"><code>vignettes/a_discovery_and_retrieval.Rmd</code></a></small>
      <div class="d-none name"><code>a_discovery_and_retrieval.Rmd</code></div>
    </div>

    
    
<p>CxGcensus is an alternative <em>R</em> client to the <a href="https://chanzuckerberg.github.io/cellxgene-census/" class="external-link">CELLxGENE
census</a>. It emaphsizes use cases related to data discovery, and uses
in-memory and on-disk caches to reduce latency associated with repeated
queries.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install CxGcensus from <a href="https://github.com/" class="external-link">GitHub</a>
with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"mtmorgan/CxGcensus"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-exploration">Data exploration<a class="anchor" aria-label="anchor" href="#data-exploration"></a>
</h2>
<p>Load the package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mtmorgan.github.io/CxGcensus/">CxGcensus</a></span><span class="op">)</span></span></code></pre></div>
<p>This is the census and package versions we are using</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/census.html">census</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">get_metadata</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">as.vector</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ created_on           : chr "2023-07-25T13:50:53+00:00"</span></span>
<span><span class="co">#&gt;  $ git_commit_sha       : chr "75c2fc7"</span></span>
<span><span class="co">#&gt;  $ soma_encoding_version: chr "1"</span></span>
<span><span class="co">#&gt;  $ soma_object_type     : chr "SOMACollection"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html" class="external-link">packageVersion</a></span><span class="op">(</span><span class="st">"CxGcensus"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] '0.0.0.9025'</span></span>
<span><span class="fu">tiledbsoma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tiledbsoma/man/show_package_versions.html" class="external-link">show_package_versions</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; tiledbsoma:    0.99.1</span></span>
<span><span class="co">#&gt; tiledb-r:      0.20.1</span></span>
<span><span class="co">#&gt; tiledb core:   2.16.0</span></span>
<span><span class="co">#&gt; libtiledbsoma: libtiledb=2.16.0</span></span>
<span><span class="co">#&gt; R:             R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; OS:            Ubuntu 22.04.2 LTS</span></span></code></pre></div>
<p>Discover datasets used to construct the census</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/census_info.html">datasets</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 593 × 8</span></span></span>
<span><span class="co">#&gt;    soma_joinid collection_id           collection_name collection_doi dataset_id</span></span>
<span><span class="co">#&gt;          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>           0 e2c257e7-6f79-487c-b81… Spatial multio… 10.1038/s4158… f171db61-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>           1 e2c257e7-6f79-487c-b81… Spatial multio… 10.1038/s4158… ecf2e08e-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>           2 e2c257e7-6f79-487c-b81… Spatial multio… 10.1038/s4158… 74cff64f-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>           3 f7cecffa-00b4-4560-a29… Mapping single… 10.1016/j.cce… 5af90777-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>           4 3f50314f-bdc9-40c6-8e4… Single-cell se… 10.1016/j.cce… bd65a70f-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>           5 e5f58829-1a66-40b5-a62… Tabula Sapiens  10.1126/scien… ff45e623-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>           6 e5f58829-1a66-40b5-a62… Tabula Sapiens  10.1126/scien… f01bdd17-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>           7 e5f58829-1a66-40b5-a62… Tabula Sapiens  10.1126/scien… e6a11140-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>           8 e5f58829-1a66-40b5-a62… Tabula Sapiens  10.1126/scien… e5c63d94-…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>           9 e5f58829-1a66-40b5-a62… Tabula Sapiens  10.1126/scien… d8732da6-…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 583 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: dataset_title &lt;chr&gt;, dataset_h5ad_path &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   dataset_total_cell_count &lt;int&gt;</span></span></span></code></pre></div>
<p><code>dataset_id</code> and <code>collection_id</code> can be used in
the <a href="https://bioconductor.org/packages/cellxgenedp" class="external-link">cellxgenedp</a>
package for programmatic retrieval and visualization of individual
datasets.</p>
<p>The organisms present in the census (use these in calls to
<code><a href="../reference/census_data.html">feature_data()</a></code> or <code><a href="../reference/census_data.html">observation_data()</a></code>) are
available with</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/census.html">census_names</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "mus_musculus" "homo_sapiens"</span></span></code></pre></div>
<p>Summarize information about cells in the census</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/census_info.html">summary_cell_counts</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">category</span> <span class="op">==</span> <span class="st">"sex"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>        <span class="va">organism</span>, <span class="va">label</span>,</span>
<span>        <span class="va">unique_cell_count</span>, <span class="va">total_cell_count</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   organism     label   unique_cell_count total_cell_count</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Homo sapiens female           13<span style="text-decoration: underline;">955</span>052         22<span style="text-decoration: underline;">198</span>837</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Homo sapiens male             17<span style="text-decoration: underline;">152</span>628         30<span style="text-decoration: underline;">950</span>702</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Homo sapiens unknown           2<span style="text-decoration: underline;">256</span>562          3<span style="text-decoration: underline;">251</span>334</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Mus musculus female            1<span style="text-decoration: underline;">467</span>473          1<span style="text-decoration: underline;">832</span>115</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Mus musculus male              2<span style="text-decoration: underline;">423</span>438          3<span style="text-decoration: underline;">230</span>510</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Mus musculus unknown            <span style="text-decoration: underline;">192</span>620           <span style="text-decoration: underline;">192</span>620</span></span></code></pre></div>
<p>Learn about features (genes) in, e.g., <code>homo_sapiens</code>
datasets in the census</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/census_data.html">feature_data</a></span><span class="op">(</span><span class="st">"mus_musculus"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 52,392 × 4</span></span></span>
<span><span class="co">#&gt;    soma_joinid feature_id         feature_name feature_length</span></span>
<span><span class="co">#&gt;          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>           0 ENSMUSG00000051951 Xkr4                   <span style="text-decoration: underline;">6</span>094</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>           1 ENSMUSG00000089699 Gm1992                  250</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>           2 ENSMUSG00000102343 Gm37381                <span style="text-decoration: underline;">1</span>364</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>           3 ENSMUSG00000025900 Rp1                   <span style="text-decoration: underline;">12</span>311</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>           4 ENSMUSG00000025902 Sox17                  <span style="text-decoration: underline;">4</span>772</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>           5 ENSMUSG00000104328 Gm37323                <span style="text-decoration: underline;">2</span>773</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>           6 ENSMUSG00000033845 Mrpl15                 <span style="text-decoration: underline;">8</span>487</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>           7 ENSMUSG00000025903 Lypla1                 <span style="text-decoration: underline;">7</span>145</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>           8 ENSMUSG00000104217 Gm37988                 825</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>           9 ENSMUSG00000033813 Tcea1                  <span style="text-decoration: underline;">3</span>017</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 52,382 more rows</span></span></span></code></pre></div>
<p>Observation (cell) data annotations are available with
<code><a href="../reference/census_data.html">observation_data()</a></code>. The strategy adopted is to download all
the data to a local <a href="https://cran.r-project.org/package=duckdb" class="external-link">duckdb</a> database.
Download can be time-consuming, but the results are ‘memoised’ so
subsequent calls are more-or-less instantaneous.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/census_data.html">observation_data</a></span><span class="op">(</span><span class="st">"mus_musculus"</span><span class="op">)</span></span>
<span><span class="va">mus</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Source:   table&lt;obs&gt; [?? x 21]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Database: DuckDB 0.8.1 [unknown@Linux 5.15.0-1041-azure:R 4.3.1//home/runner/.cache/R/CxGcensus/80af4af4750a.duckdb]</span></span></span>
<span><span class="co">#&gt;    soma_joinid dataset_id                 assay assay_ontology_term_id cell_type</span></span>
<span><span class="co">#&gt;        <span style="color: #949494; font-style: italic;">&lt;int64&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>           0 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L5/6 nea…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>           1 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L4/5 int…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>           2 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L2/3-6 i…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>           3 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            hippocam…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>           4 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L6 corti…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>           5 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L4/5 int…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>           6 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L6b glut…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>           7 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L4/5 int…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>           8 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L4/5 int…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>           9 d7291f04-fbbb-4d65-990a-f… 10x … EFO:0009899            L6b glut…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 16 more variables: cell_type_ontology_term_id &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   development_stage &lt;chr&gt;, development_stage_ontology_term_id &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   disease &lt;chr&gt;, disease_ontology_term_id &lt;chr&gt;, donor_id &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   is_primary_data &lt;lgl&gt;, self_reported_ethnicity &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   self_reported_ethnicity_ontology_term_id &lt;chr&gt;, sex &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sex_ontology_term_id &lt;chr&gt;, suspension_type &lt;chr&gt;, tissue &lt;chr&gt;, …</span></span></span></code></pre></div>
<p>The columns available for cross-dataset analysis are</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mus</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "soma_joinid"                             </span></span>
<span><span class="co">#&gt;  [2] "dataset_id"                              </span></span>
<span><span class="co">#&gt;  [3] "assay"                                   </span></span>
<span><span class="co">#&gt;  [4] "assay_ontology_term_id"                  </span></span>
<span><span class="co">#&gt;  [5] "cell_type"                               </span></span>
<span><span class="co">#&gt;  [6] "cell_type_ontology_term_id"              </span></span>
<span><span class="co">#&gt;  [7] "development_stage"                       </span></span>
<span><span class="co">#&gt;  [8] "development_stage_ontology_term_id"      </span></span>
<span><span class="co">#&gt;  [9] "disease"                                 </span></span>
<span><span class="co">#&gt; [10] "disease_ontology_term_id"                </span></span>
<span><span class="co">#&gt; [11] "donor_id"                                </span></span>
<span><span class="co">#&gt; [12] "is_primary_data"                         </span></span>
<span><span class="co">#&gt; [13] "self_reported_ethnicity"                 </span></span>
<span><span class="co">#&gt; [14] "self_reported_ethnicity_ontology_term_id"</span></span>
<span><span class="co">#&gt; [15] "sex"                                     </span></span>
<span><span class="co">#&gt; [16] "sex_ontology_term_id"                    </span></span>
<span><span class="co">#&gt; [17] "suspension_type"                         </span></span>
<span><span class="co">#&gt; [18] "tissue"                                  </span></span>
<span><span class="co">#&gt; [19] "tissue_ontology_term_id"                 </span></span>
<span><span class="co">#&gt; [20] "tissue_general"                          </span></span>
<span><span class="co">#&gt; [21] "tissue_general_ontology_term_id"</span></span></code></pre></div>
<p>It is straight-forward to summarize or query the data, e.g., for the
number of different types of assays represented, or for studies of
diabetes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mus</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">assay</span>, sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Source:     SQL [9 x 2]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Database:   DuckDB 0.8.1 [unknown@Linux 5.15.0-1041-azure:R 4.3.1//home/runner/.cache/R/CxGcensus/80af4af4750a.duckdb]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Ordered by: desc(n)</span></span></span>
<span><span class="co">#&gt;   assay                                n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                          <span style="color: #949494; font-style: italic;">&lt;int64&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 10x 3' v2                      2<span style="text-decoration: underline;">815</span>560</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 10x 3' v3                      1<span style="text-decoration: underline;">902</span>339</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Smart-seq2                      <span style="text-decoration: underline;">221</span>648</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 10x 5' v1                       <span style="text-decoration: underline;">161</span>525</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Drop-seq                         <span style="text-decoration: underline;">80</span>990</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Smart-seq                        <span style="text-decoration: underline;">48</span>972</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> 10x 3' v1                        <span style="text-decoration: underline;">21</span>421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> DroNc-seq                         <span style="text-decoration: underline;">2</span>768</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> 10x 3' transcription profiling      22</span></span>
<span><span class="va">mus</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"diabetes"</span>, <span class="va">disease</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">disease</span>, <span class="va">sex</span>, <span class="va">tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Source:   SQL [2 x 4]</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Database: DuckDB 0.8.1 [unknown@Linux 5.15.0-1041-azure:R 4.3.1//home/runner/.cache/R/CxGcensus/80af4af4750a.duckdb]</span></span></span>
<span><span class="co">#&gt;   disease                  sex    tissue                    n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;int64&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> type 2 diabetes mellitus male   islet of Langerhans   <span style="text-decoration: underline;">99</span>747</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> type 1 diabetes mellitus female islet of Langerhans   <span style="text-decoration: underline;">39</span>932</span></span></code></pre></div>
<p>Use <a href="https://cran.r-project.org/package=ggplot2" class="external-link">ggplot2</a>
to visualize tissue type and assay.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 10 most-common tissues; total cell count</span></span>
<span><span class="va">common_tissues</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">mus</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">tissue</span>, sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span></span>
<span><span class="co">## assays on each tissue</span></span>
<span><span class="va">tissue_and_assay</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">mus</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">tissue</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">common_tissues</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="va">assay</span>, sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## visualize</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tissue_and_assay</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">tissue</span>, levels <span class="op">=</span> <span class="va">common_tissues</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">n</span>, fill <span class="op">=</span> <span class="va">assay</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>position<span class="op">=</span><span class="st">"stack"</span>, stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Tissue"</span>, y <span class="op">=</span> <span class="st">"Total cell count"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.8</span>, <span class="fl">.7</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="a_discovery_and_retrieval_files/figure-html/mus-tissue-assay-1.png" width="700"></p>
<p>The <code>soma_joinid</code> in the tibbles returned by
<code><a href="../reference/census_data.html">feature_data()</a></code> and <code><a href="../reference/census_data.html">observation_data()</a></code> are
important in retrieving counts.</p>
</div>
<div class="section level2">
<h2 id="experiment-retrieval-as-singlecellexperiment">Experiment retrieval as SingleCellExperiment<a class="anchor" aria-label="anchor" href="#experiment-retrieval-as-singlecellexperiment"></a>
</h2>
<p>To create a Bioconductor <a href="https://bioconductor.org/packages/SingleCellExperiment" class="external-link">SingleCellExperiment</a>,
perhaps for use in one of the <a href="https://bioconductor.org/books/OSCA" class="external-link">Orchestrating Single Cell
Analysis with Bioconductor</a> (OSCA) workflows, first select desired
features and observations. For illustration we choose 100 random
genes</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12</span><span class="op">)</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/census_data.html">feature_data</a></span><span class="op">(</span><span class="st">"mus_musculus"</span><span class="op">)</span></span>
<span><span class="va">random_features</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">features</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>and choose all cells from brain tissue and development stage 20 month
old or latter.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">observations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/census_data.html">observation_data</a></span><span class="op">(</span><span class="st">"mus_musculus"</span><span class="op">)</span></span>
<span><span class="va">brain_20mo</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">observations</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"brain"</span>,</span>
<span>        <span class="va">development_stage</span> <span class="op">==</span> <span class="st">"20 month-old stage and over"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Cells are from several datasets</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">brain_20mo</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 1</span></span></span>
<span><span class="co">#&gt;       n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> <span style="text-decoration: underline;">52</span>132</span></span>
<span><span class="va">brain_20mo</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">dataset_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 2</span></span></span>
<span><span class="co">#&gt;   dataset_id                               n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 3bbb6cf9-72b9-41be-b568-656de6eb18b5 <span style="text-decoration: underline;">38</span>695</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 58b01044-c5e5-4b0f-8a2d-6ebf951e01ff  <span style="text-decoration: underline;">3</span>077</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 66ff82b4-9380-469c-bc4b-cfa08eacd325   756</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 98e5ea9f-16d6-47ec-a529-686e76515e39  <span style="text-decoration: underline;">5</span>180</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> c08f8441-4a10-4748-872a-e70c0bcccdba  <span style="text-decoration: underline;">4</span>424</span></span></code></pre></div>
<p>Use these to create a <code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/single_cell_experiment.html">single_cell_experiment</a></span><span class="op">(</span><span class="st">"mus_musculus"</span>, <span class="va">random_features</span>, <span class="va">brain_20mo</span><span class="op">)</span></span>
<span><span class="va">sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 100 52132 </span></span>
<span><span class="co">#&gt; metadata(1): census_metadata</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(4): soma_joinid feature_id feature_name feature_length</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(21): soma_joinid dataset_id ... tissue_general</span></span>
<span><span class="co">#&gt;   tissue_general_ontology_term_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>The experiment has <code><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts()</a></code> as a sparse
(<code>dgCMatrix</code>, from the Matrix package) matrix, with
<code>features</code> and <code>observations</code> available (as
<em>Bioconductor</em> <code>DataFrame</code> objects, rather than
tibbles) via <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>. The data
is straight-forward to work with, e.g., removing features for which no
counts were observed…</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="va">keep_rows</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span>, withDimnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">keep_rows</span><span class="op">)</span></span>
<span><span class="co">#&gt; keep_rows</span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;    58    42</span></span>
<span><span class="va">sce</span><span class="op">[</span><span class="va">keep_rows</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 42 52132 </span></span>
<span><span class="co">#&gt; metadata(1): census_metadata</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames: NULL</span></span>
<span><span class="co">#&gt; rowData names(4): soma_joinid feature_id feature_name feature_length</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(21): soma_joinid dataset_id ... tissue_general</span></span>
<span><span class="co">#&gt;   tissue_general_ontology_term_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>… or summarizing the total number of reads observed in each cell</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span>, withDimnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Counts per cell"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"log(1 + x) counts"</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a_discovery_and_retrieval_files/figure-html/sce-colSums-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This article was compiled with CxGcensus version 0.0.0.9025. Full
session info is:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.2</span></span>
<span><span class="co">#&gt;  [3] Biobase_2.60.0              GenomicRanges_1.52.0       </span></span>
<span><span class="co">#&gt;  [5] GenomeInfoDb_1.36.1         IRanges_2.34.1             </span></span>
<span><span class="co">#&gt;  [7] S4Vectors_0.38.1            BiocGenerics_0.46.0        </span></span>
<span><span class="co">#&gt;  [9] MatrixGenerics_1.12.3       matrixStats_1.0.0          </span></span>
<span><span class="co">#&gt; [11] ggplot2_3.4.2               CxGcensus_0.0.0.9025       </span></span>
<span><span class="co">#&gt; [13] RcppSpdlog_0.0.14           dplyr_1.1.2                </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0            farver_2.1.1               </span></span>
<span><span class="co">#&gt;  [3] blob_1.2.4                  bitops_1.0-7               </span></span>
<span><span class="co">#&gt;  [5] urltools_1.7.3              arrow_12.0.1.1             </span></span>
<span><span class="co">#&gt;  [7] RCurl_1.98-1.12             fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [9] duckdb_0.8.1-1              digest_0.6.33              </span></span>
<span><span class="co">#&gt; [11] lifecycle_1.0.3             magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [13] compiler_4.3.1              rlang_1.1.1                </span></span>
<span><span class="co">#&gt; [15] sass_0.4.7                  tools_4.3.1                </span></span>
<span><span class="co">#&gt; [17] utf8_1.2.3                  yaml_2.3.7                 </span></span>
<span><span class="co">#&gt; [19] data.table_1.14.8           knitr_1.43                 </span></span>
<span><span class="co">#&gt; [21] S4Arrays_1.0.5              labeling_0.4.2             </span></span>
<span><span class="co">#&gt; [23] bit_4.0.5                   spdl_0.0.5                 </span></span>
<span><span class="co">#&gt; [25] curl_5.0.1                  DelayedArray_0.26.7        </span></span>
<span><span class="co">#&gt; [27] aws.signature_0.6.0         xml2_1.3.5                 </span></span>
<span><span class="co">#&gt; [29] abind_1.4-5                 withr_2.5.0                </span></span>
<span><span class="co">#&gt; [31] purrr_1.0.1                 desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [33] triebeard_0.4.1             grid_4.3.1                 </span></span>
<span><span class="co">#&gt; [35] fansi_1.0.4                 colorspace_2.1-0           </span></span>
<span><span class="co">#&gt; [37] scales_1.2.1                cli_3.6.1                  </span></span>
<span><span class="co">#&gt; [39] crayon_1.5.2                rmarkdown_2.23             </span></span>
<span><span class="co">#&gt; [41] ragg_1.2.5                  generics_0.1.3             </span></span>
<span><span class="co">#&gt; [43] tiledbsoma_0.99.1           httr_1.4.6                 </span></span>
<span><span class="co">#&gt; [45] DBI_1.1.3                   cachem_1.0.8               </span></span>
<span><span class="co">#&gt; [47] stringr_1.5.0               zlibbioc_1.46.0            </span></span>
<span><span class="co">#&gt; [49] assertthat_0.2.1            XVector_0.40.0             </span></span>
<span><span class="co">#&gt; [51] base64enc_0.1-3             vctrs_0.6.3                </span></span>
<span><span class="co">#&gt; [53] Matrix_1.5-4.1              jsonlite_1.8.7             </span></span>
<span><span class="co">#&gt; [55] bit64_4.0.5                 systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt; [57] jquerylib_0.1.4             glue_1.6.2                 </span></span>
<span><span class="co">#&gt; [59] pkgdown_2.0.7               stringi_1.7.12             </span></span>
<span><span class="co">#&gt; [61] gtable_0.3.3                aws.s3_0.3.21              </span></span>
<span><span class="co">#&gt; [63] munsell_0.5.0               nanotime_0.3.7             </span></span>
<span><span class="co">#&gt; [65] tibble_3.2.1                pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [67] htmltools_0.5.5             GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">#&gt; [69] R6_2.5.1                    dbplyr_2.3.3               </span></span>
<span><span class="co">#&gt; [71] textshaping_0.3.6           rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt; [73] evaluate_0.21               lattice_0.21-8             </span></span>
<span><span class="co">#&gt; [75] RcppCCTZ_0.2.12             highr_0.10                 </span></span>
<span><span class="co">#&gt; [77] cellxgene.census_0.0.0.9000 memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [79] tiledb_0.20.1               bslib_0.5.0                </span></span>
<span><span class="co">#&gt; [81] rjsoncons_1.0.0             Rcpp_1.0.11                </span></span>
<span><span class="co">#&gt; [83] xfun_0.39                   fs_1.6.3                   </span></span>
<span><span class="co">#&gt; [85] zoo_1.8-12                  pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Morgan, Chan Zuckerberg Initiative DAF CZF2019-002443.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
